source/common/upstream/health_checker_impl.cc

```diff
@@ -113,8 +109,17 @@ HttpHealthCheckerImpl::HttpHealthCheckerImpl(const Cluster& cluster,
       codec_client_type_(codecClientType(config.http_health_check().use_http2()
                                              ? envoy::type::HTTP2
                                              : config.http_health_check().codec_client_type())) {
+
+  if (config.http_health_check().has_service_name_matcher()) {
+    service_name_matcher_ = std::make_unique<Matchers::StringMatcherImpl>(
+        config.http_health_check().service_name_matcher());
+  }
   if (!config.http_health_check().service_name().empty()) {
-    service_name_ = config.http_health_check().service_name();
+    envoy::api::v2::core::HealthCheck mutable_config = config;
+    mutable_config.mutable_http_health_check()->mutable_service_name_matcher()->set_prefix(
+        config.http_health_check().service_name());
+    service_name_matcher_ = std::make_unique<Matchers::StringMatcherImpl>(
+        mutable_config.http_health_check().service_name_matcher());
```

Comment:
@mattklein123 I am not sure if this is the best way to create a proto object. I would like to improve this part if there's a better way of doing it.

Reply:
Is this a better way?
```
envoy::api::v2::core::HealthCheck config_(config);
    config_.mutable_http_health_check()->mutable_service_name_matcher()->set_prefix(
        config.http_health_check().service_name());
    service_name_matcher_.emplace(config_.http_health_check().service_name_matcher());
```